File: app/src/main/java/com/yawar/memo/adapter/DeviceLinkAdapter.java
package com.yawar.memo.adapter;

import android.app.Activity;
import android.content.ContentProviderOperation;
import android.content.Context;
import android.content.SharedPreferences;
import android.database.Cursor;
import android.net.Uri;
import android.provider.ContactsContract;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.Glide;
import com.bumptech.glide.request.RequestOptions;
import com.yawar.memo.R;
import com.yawar.memo.constant.AllConstants;
import com.yawar.memo.fragment.SearchFragment;
import com.yawar.memo.model.DeviceLinkModel;
import com.yawar.memo.model.SearchRespone;

import java.util.ArrayList;


public class DeviceLinkAdapter extends RecyclerView.Adapter<com.yawar.memo.adapter.DeviceLinkAdapter.ViewHolders> {
    ///Initialize variable
    Activity activity;
    float textSize = 14.0F ;
    SharedPreferences sharedPreferences ;

    ArrayList<DeviceLinkModel> arrayList;
    private CallbackInterface mCallback;
    public interface CallbackInterface{

        /**
         * Callback invoked when clicked
         * @param position - the position
         * @param deviceLinkModel - the text to pass back
         */
        void onHandleSelection(int position, DeviceLinkModel deviceLinkModel);
        void onClickItem(int position, DeviceLinkModel deviceLinkModel);

    }


    ///Create constructor
    public DeviceLinkAdapter(Activity activity, ArrayList<DeviceLinkModel> arrayList) {
        this.activity = activity;
        this.arrayList = arrayList;
        try{
            mCallback = (CallbackInterface) activity;
        }catch(ClassCastException ex){
            //.. should log the error or throw and exception
        }
        notifyDataSetChanged();
    }


    @NonNull
    @Override
    public com.yawar.memo.adapter.DeviceLinkAdapter.ViewHolders onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        ///Initialize view
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_device_link, parent, false);


        DeviceLinkAdapter.ViewHolders holder = new DeviceLinkAdapter.ViewHolders(view);

        return holder;
    }


    @Override
    public void onBindViewHolder(@NonNull com.yawar.memo.adapter.DeviceLinkAdapter.ViewHolders holder, int position) {

        DeviceLinkModel model = arrayList.get(position);
        holder.tvName.setText(model.getName());
        holder.tvNumber.setText(model.getTime());
        System.out.println(model.getImage());

        // Glide.with(holder.imageView.getContext()).load(model.getImage()).into(holder.imageView);
        if(!model.getImage().isEmpty()){
//            Glide.with(holder.imageView.getContext()).load(AllConstants.imageUrl+model.getImage()).error(activity.getDrawable(R.drawable.th)).into(holder.imageView);
            Glide.with(holder.imageView.getContext()).load(AllConstants.imageUrl+model.getImage()).apply(RequestOptions.placeholderOf(R.drawable.th).error(R.drawable.th)).into(holder.imageView);
        }

        holder.itemView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if(mCallback != null){
                    mCallback.onClickItem(position, arrayList.get(position));
                }
            }
        });







    }


    @Override
    public int getItemCount() {
        return arrayList.size();
    }




    public class ViewHolders extends RecyclerView.ViewHolder {
        ///Initialize variable
        TextView tvName;
        ImageView imageView;
        TextView tvNumber;

        Button button;
        public ViewHolders(@NonNull View itemView) {
            super(itemView);
            ////Assign variable
            tvName = itemView.findViewById(R.id.tv_name);
                tvName.setTextSize(textSize);
                tvName.setTextSize(Float.parseFloat(sharedPreferences.getString("txtFontSize", "16")));

            tvNumber = itemView.findViewById(R.id.tv_number);
                tvNumber.setTextSize(textSize);
                tvNumber.setTextSize(Float.parseFloat(sharedPreferences.getString("txtFontSize", "16")));


            imageView = itemView.findViewById(R.id.iv_image);
            button = itemView.findViewById(R.id.btn_add);
        }
    }
}

File: app/src/main/java/com/yawar/memo/ui/deviceLinkPage/DeviceLinkAdapter.kt
package com.yawar.memo.ui.deviceLinkPage

import android.app.Activity
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.bumptech.glide.request.RequestOptions
import com.yawar.memo.R
import com.yawar.memo.constant.AllConstants
import com.yawar.memo.model.DeviceLinkModel


class DeviceLinkAdapter(
    var activity: Activity, var arrayList: ArrayList<DeviceLinkModel>,
) :
    RecyclerView.Adapter<DeviceLinkAdapter.ViewHolder>() {
    private var mCallback: CallbackInterface? = null

    interface CallbackInterface {

        fun onHandleSelection(position: Int, deviceLinkModel: DeviceLinkModel?)
        fun onClickItem(position: Int, deviceLinkModel: DeviceLinkModel?)
    }
    init {
        try {
            mCallback = activity as CallbackInterface
        } catch (ex: ClassCastException) {
        }
    }

    override fun onCreateViewHolder(
        parent: ViewGroup,
        viewType: Int,
    ): ViewHolder {
        return ViewHolder.from(parent)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val model = arrayList[position]
        holder.bind(model, position, mCallback)
    }



    override fun getItemCount(): Int {
        return arrayList.size
    }

     class ViewHolder private  constructor (itemView: View) : RecyclerView.ViewHolder(itemView) {
        ///Initialize variable
        var tvName: TextView = itemView.findViewById(R.id.tv_name)
        var imageView: ImageView = itemView.findViewById(R.id.iv_image)
        var tvNumber: TextView = itemView.findViewById(R.id.tv_number)
        var button: Button = itemView.findViewById(R.id.btn_add)



        fun bind(
            model: DeviceLinkModel,
            position: Int,
            mCallback: CallbackInterface?
        ) {
            tvName.text = model.name
            tvNumber.text = model.time
            println(model.image)

            if (model.image!!.isNotEmpty()) {
                Glide.with(imageView.context).load(AllConstants.imageUrl + model.image)
                    .apply(RequestOptions.placeholderOf(R.drawable.th).error(R.drawable.th))
                    .into(imageView)
            }
            itemView.setOnClickListener {
                mCallback?.onClickItem(position, model)
            }
        }
         companion object {
             fun from(parent: ViewGroup): ViewHolder {
                 val view: View =
                     LayoutInflater.from(parent.context).inflate(R.layout.item_device_link, parent, false)
                 return ViewHolder(view)
             }
         }


    }



}